### ISSUE TITLE
test(rust-rpc-test): have a contract to declare which is not already declared

### ISSUE THREAD

**Issue by tdelabro** [keep-starknet-strange/madara#1110]
during the genesis phase of the sharingan test-chains, we populate the state with a few contracts already declared and deployed.

One of them is ERC20.json. Due to some shenanigans, it was declared with another arbitrary class hash, making it possible for users to re-declare it a second time, this time with it's real class hash.

Some tests used this contract in order to test the declare workflow `starknet-rpc-test::get_transaction_receipt:: work_with_declare_transaction` is a survivant of this era that we are ignoring for now.

One should activate this test again by making it declare another contract class, one which is not already present in the storage

**Comment by NaitikChaudhary** [comment]
Can I take this one?

**Comment by apoorvsadana** [comment]
Assigned to you!

**Comment by tdelabro** [comment]
Hey, all good?
We can unassign you if you can't find time. No big deal

**Comment by NaitikChaudhary** [comment]
Hey, last few weeks were really hectic for me couldn’t take out time, but
this week I can take out time for sure, so is it fine if I create a PR by
this weekend?

**Comment by tdelabro** [comment]
Yeah sure.
2 ways you can do it.
- group all the tests regarding declare in one single tests and do all the checks after a single declare 
- create n contracts with one internal constant ranging from 0 to n. Then use one of them in each test that need to call declare

**Comment by tdelabro** [comment]
@NaitikChaudhary  All good?

**Comment by tdelabro** [comment]
@NaitikChaudhary Will you deliver this one or can I unassign it from you?

**Comment by apoorvsadana** [comment]
I don't think he's working on it. Unassigned him.

**Comment by edisontim** [comment]
Hey, can I take this :) ?

**Comment by tdelabro** [comment]
@edisontim everything for my fellow 42 alumni

**Comment by edisontim** [comment]
Thanks! On it :)

**Comment by edisontim** [comment]
> Yeah sure. 2 ways you can do it.
> 
> * group all the tests regarding declare in one single tests and do all the checks after a single declare
> * create n contracts with one internal constant ranging from 0 to n. Then use one of them in each test that need to call declare

Contract being declared in ```starknet-rpc-test::get_transaction_receipt:: work_with_declare_transaction``` is ```Counter.json```, so this is the contract that can't be redeclared multiple times right? 

I think I'll go first the first solution, this seems the cleanest now. However, if more declare tests need to happen this could get messy

**Comment by tdelabro** [comment]
Yes that is the one.

I think first option is okay. Yes you will have to organize and comment your code correctly so it don't get nightmarish

**Comment by edisontim** [comment]
Had a look at a global variable shared between all tokio tests or a single use init function that launched before tests but wasn't able to find anything viable :/

**Comment by tdelabro** [comment]
Wdym?
You have to run your node manually and then launch the test.

**Comment by edisontim** [comment]
I mean having a setup function before all rpc-tests that would declare the contract once for all tests and make available a global declare_tx variable, but it seems like with async tests that's impossible

**Comment by tdelabro** [comment]
Yep, there is no `init` logic in cargo test. I would not rely on this behavior.
What you can do is create a function `deploy_if_not_already() { if already_declared {} else { declare() } }` and call it in all your tests that need this contract deployed. A bit hacky but it solve you problem

**Comment by edisontim** [comment]
Yeah I'm having a look at that now, it's cleaner than having n contracts declared imo

**Comment by edisontim** [comment]
Would you mind having a look at telling me if I'm wrong? I think what I'm trying to do is not possible:
I'm managing to get a global ref set at initialisation containing the return of ```account.declare_tx```, but cannot use the value of the reference as the ```TransactionDeclaration``` doesn't implement the ```Clone``` trait. I'm about to drop it and just for the grouping of tests in one function.
What I have:
```
lazy_static! {
    pub static ref _MADARA_CLIENT: MadaraClientThreadSafe = MadaraClientThreadSafe::new();
    pub static ref RPC: JsonRpcClient<HttpTransport> = _MADARA_CLIENT.get_starknet_client();
    pub static ref ACCOUNT: RpcAccount<'static> =
        build_single_owner_account(&RPC, SIGNER_PRIVATE, ARGENT_CONTRACT_ADDRESS, true);
    pub static ref DECLARE_TX_TUP: (TransactionDeclaration<'static>, FieldElement, FieldElement) =
        ACCOUNT.declare_contract("./contracts/Counter.sierra.json", "./contracts/Counter.casm.json");
}
```
```
        let txs = madara_write_lock.create_block_with_txs(vec![Transaction::Declaration(DECLARE_TX_TUP.0)]).await?;
```
compiler error: 
```
error[E0507]: cannot move out of dereference of `DECLARE_TX_TUP`
   --> starknet-rpc-test/add_declare_transaction.rs:100:89
    |
100 |         let txs = madara_write_lock.create_block_with_txs(vec![Transaction::Declaration(DECLARE_TX_TUP.0)]).await?;
    |                                                                                         ^^^^^^^^^^^^^^^^ move occurs because value has type `starknet_accounts::Declaration<'_, SingleOwnerAccount<&JsonRpcClient<HttpTransport>, starknet_signers::local_wallet::LocalWallet>>`, which does not implement the `Copy` trait```
